src/lua/protobuf_wireformat.lua

```diff
@@ -0,0 +1,98 @@
+local ffi = require('ffi')
+
+-- Internal function for protobuf integer encoding.
+-- Used for VARINT encoding and tag encoding for other types.
+local function internal_encode_uint(data)
+    local code = ''
+    local byte
+    data = ffi.cast('uint64_t', data)
+    repeat
+        byte = (data % 128) + 128
+        data = data / 128
+        if data == 0 then
+            byte = byte - 128
+        end
+        code = code .. string.char(tonumber(byte))
+    until data == 0
+    return code
+end
+
+-- Encode for any VARINT values except sintN
+local function encode_uint(data, field_id)
+    local wire_type = 0
+    return internal_encode_uint(bit.lshift(field_id, 3) + wire_type) ..
+        internal_encode_uint(data)
+end
+
+-- "ZigZag" encoding for sintN values in VARINT
+local function encode_sint(data, field_id)
+    if data > 0 then
+        return encode_uint(2 * data, field_id)
+    else
+        return encode_uint(2 * (-data) - 1, field_id)
+    end
+end
```

Comment:
The `data > 0` check is incorrect. We should encode zero as `tag + varint(0)`, but this code gives `tag + varint(-1)`.

The check must be `data >= 0`. Please, add test cases for `sint32` and `sint64` with a zero value.

Moreover, `sint32` and `sint64` encoding is incorrect now: it encodes `tag + varint(data)` instead of `tag + varint(data * 2)` for `data >= 0`.

It seems it should be fixed this way:

```diff
diff --git a/src/lua/protobuf.lua b/src/lua/protobuf.lua
index f15d0bc76b..684a076ac5 100644
--- a/src/lua/protobuf.lua
+++ b/src/lua/protobuf.lua
@@ -470,13 +470,7 @@ scalars.sint32 = {
         check_integer(field_def, value)
         validate_value(field_def, value, MIN_SINT32, MAX_SINT32)
     end,
-    encode = function(value, field_id)
-        if value >= 0 then
-            return wireformat.encode_uint(value, field_id)
-        elseif value < 0 then
-            return wireformat.encode_sint(value, field_id)
-        end
-    end,
+    encode = wireformat.encode_sint,
 }
 
 scalars.uint32 = {
@@ -541,13 +535,7 @@ scalars.sint64 = {
             validate_value(field_def, value, MIN_CDATA, MAX_CDATA)
         end
     end,
-    encode = function(value, field_id)
-        if value >= 0 then
-            return wireformat.encode_uint(value, field_id)
-        elseif value < 0 then
-            return wireformat.encode_sint(value, field_id)
-        end
-    end,
+    encode = wireformat.encode_sint,
 }
```

It fails 4 test cases in `app-luatest/protobuf_varint_test.lua`. I guess that the test cases are incorrect and I think we should verify the test cases and fix them.